### PR TITLE
[SPARK-44876][PYTHON] Fix Arrow-optimized Python UDF on Spark Connect

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Fixes Arrow-optimized Python UDF on Spark Connect.

Also enables the missing test `pyspark.sql.tests.connect.test_parity_arrow_python_udf`.

### Why are the changes needed?

`pyspark.sql.tests.connect.test_parity_arrow_python_udf` is not listed in `dev/sparktestsupport/modules.py`, and it fails when running manually.

```
======================================================================
ERROR [0.072s]: test_register (pyspark.sql.tests.connect.test_parity_arrow_python_udf.ArrowPythonUDFParityTests)
----------------------------------------------------------------------
Traceback (most recent call last):
...
pyspark.errors.exceptions.base.PySparkRuntimeError: [SCHEMA_MISMATCH_FOR_PANDAS_UDF] Result vector from pandas_udf was not the required length: expected 1, got 38.
```

The failure had not been captured because the test is missing in the `module.py` file.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @xinrong-meng @HyukjinKwon

File: python/pyspark/sql/udf.py [hunk 1]
```diff
@@ -131,58 +130,24 @@ def _create_py_udf(
...
+
     if is_arrow_enabled:
+        try:
+            is_func_with_args = len(getfullargspec(f).args) > 0
+        except TypeError:
+            is_func_with_args = False
         if is_func_with_args:
-            return _create_arrow_py_udf(regular_udf)
+            require_minimum_pandas_version()
+            require_minimum_pyarrow_version()
```
Comment (ueshin) [inline python/pyspark/sql/udf.py:142]: @xinrong-meng Do we want to fail here? or fallback to non-Arrow similar to Python UDTF?

https://github.com/apache/spark/blob/5526efc206e7f98f73a141151477a5022efff05e/python/pyspark/sql/udtf.py#L116-L127

cc @allisonwang-db

Comment (xinrong-meng) [reply python/pyspark/sql/udf.py:142]: I would suggest for Arrow Python UDF, we fail in Spark 3.5. And when it's turned on by default in Spark 4.0, we fall back. WDYT?

**Review by xinrong-meng — APPROVED** [review]

File: python/pyspark/sql/connect/udf.py [hunk 2]
```diff
@@ -74,22 +72,22 @@ def _create_py_udf(
     else:
         is_arrow_enabled = useArrow
 
-    regular_udf = _create_udf(f, returnType, PythonEvalType.SQL_BATCHED_UDF)
-    try:
-        is_func_with_args = len(getfullargspec(f).args) > 0
-    except TypeError:
-        is_func_with_args = False
+    eva_type: int = PythonEvalType.SQL_BATCHED_UDF
```
Comment (xinrong-meng) [inline python/pyspark/sql/connect/udf.py:75]: Just wanted to confirm if `eva_type` is typo or intentional : )

**Comment by xinrong-meng** [comment]
LGTM, pending test

File: python/pyspark/sql/udf.py (hunk 1 above)
Comment (ueshin) [reply python/pyspark/sql/udf.py:142]: Sure, then I'll just keep it as-is. Thanks.

File: python/pyspark/sql/connect/udf.py (hunk 2 above)
Comment (ueshin) [reply python/pyspark/sql/connect/udf.py:75]: It's definitely a typo. :)

**Review by dongjoon-hyun — COMMENTED** [review]
Sorry, but the PR title looks misleading to me. This PR technically implements UDF feature instead of simply enabling UDF test code. Can we have a more intuitive PR title?

**Comment by dongjoon-hyun** [comment]
To the reviewers,
- This is filed as a blocker issue for Apache Spark 3.5.0.
- However, this looks like a missing major feature which didn't implemented yet before the feature freeze.

I'm not disagree with backporting this PR, but I believe we need to give a correct PR title instead of saying a missing test coverage.

**Comment by ueshin** [comment]
@dongjoon-hyun Thanks for reviewing this!

Sure, I'll update the title and description, but as for the backport, actually this is already implemented in 3.5 at https://github.com/apache/spark/pull/39384 and https://github.com/apache/spark/pull/40725. Unfortunately it had a bug and we didn't notice it because the test has not been activated.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.5

**Pull request closed without merging**